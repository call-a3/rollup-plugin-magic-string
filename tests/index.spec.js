import { rollup } from 'rollup'
import hypothetical from 'rollup-plugin-hypothetical'

import * as rollupPluginMagicString from '..'

test('should throw if mandatory option "magic" is missing or is not a function', () => {
  expect(() =>
    rollupPluginMagicString.default({}),
  ).toThrowErrorMatchingSnapshot()
  expect(() =>
    rollupPluginMagicString.default({ magic: true }),
  ).toThrowErrorMatchingSnapshot()
})

test('should execute custom magic-string operations correctly', async () => {
  const bundle = await rollup({
    input: './test/transform.js',
    plugins: [
      hypothetical({
        files: {
          './test/transform.js': `import template from './template.html'

export default {
  template,
}
`,
          './test/template.html': `<div class="test">
  <%- msg %>
</div>
`,
        },
      }),
      rollupPluginMagicString.default({
        magic(code, _id, magicString) {
          const wrappedCode = JSON.stringify(code)
          magicString
            .overwrite(0, code.length, wrappedCode)
            .prependLeft(0, 'export default ')
        },
        include: '**/*.html',
      }),
    ],
  })

  const { code, map } = await bundle.generate({ format: 'es', sourcemap: true })

  expect(code).toMatchSnapshot()
  expect(map).not.toBeFalsy()
})

test('should append code correctly', async () => {
  const bundle = await rollup({
    input: './test/append.js',
    plugins: [
      hypothetical({
        files: {
          './test/append.js': `const append = {}
`,
        },
      }),
      rollupPluginMagicString.append('export default append'),
    ],
  })

  const { code, map } = await bundle.generate({ format: 'es', sourcemap: true })

  expect(code).toMatchSnapshot()
  expect(map).not.toBeFalsy()
})

test('should prepend code correctly', async () => {
  const bundle = await rollup({
    input: './test/prepend.js',
    plugins: [
      hypothetical({
        files: {
          './test/prepend.js': `export default prepend()
`,
        },
      }),
      rollupPluginMagicString.prepend('const prepend = () => {}\n'),
    ],
  })

  const { code, map } = await bundle.generate({ format: 'es', sourcemap: true })

  expect(code).toMatchSnapshot()
  expect(map).not.toBeFalsy()
})

test('should wrap code correctly', async () => {
  const bundle = await rollup({
    input: './test/wrap.js',
    plugins: [
      hypothetical({
        files: {
          './test/wrap.js': `const wrapped = wrap()
`,
        },
      }),
      rollupPluginMagicString.wrap(
        'const wrap = () => {}\n',
        'export default wrapped',
      ),
    ],
  })

  const { code, map } = await bundle.generate({ format: 'es', sourcemap: true })

  expect(code).toMatchSnapshot()
  expect(map).not.toBeFalsy()
})
test('should not output a sourcemap when disabled', async () => {
  const bundle = await rollup({
    input: './test/wrap.js',
    plugins: [
      hypothetical({
        files: {
          './test/wrap.js': `const wrapped = wrap()
`,
        },
      }),
      rollupPluginMagicString.wrap('(function(){', '})()', {
        sourcemap: false,
      }),
    ],
  })

  const { map } = await bundle.generate({ format: 'es', sourcemap: false })

  expect(map).toBeFalsy()
})
