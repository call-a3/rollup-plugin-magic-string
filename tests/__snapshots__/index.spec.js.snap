// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should append code correctly 1`] = `
"const append = {};

export default append;
"
`;

exports[`should execute custom magic-string operations correctly 1`] = `
"var template = \\"<!--/home/adriaan/Projects/rollup-plugin-insert/test/template.html-->\\\\n<div class=\\\\\\"test\\\\\\">\\\\n  <%- msg %>\\\\n</div>\\\\n\\";

var transform = {
  template,
};

export default transform;
"
`;

exports[`should prepend code correctly 1`] = `
"const prepend = () => {};
var prepend$1 = prepend();

export default prepend$1;
"
`;

exports[`should throw if mandatory option "magic" is missing or is not a function 1`] = `"options.magic is required"`;

exports[`should throw if mandatory option "magic" is missing or is not a function 2`] = `"options.magic is required"`;

exports[`should wrap code correctly 1`] = `
"const wrap = () => {};
const wrapped = wrap();

export default wrapped;
"
`;
